include/dlaf/matrix/distribution.h

```diff
@@ -325,11 +341,43 @@ class Distribution {
   /// Returns the global element distance between tiles along the @p rc coordinate
   template <Coord rc>
   SizeType globalTileElementDistance(SizeType i_begin, SizeType i_end) const noexcept {
-    DLAF_ASSERT(i_begin <= i_end, i_begin, i_end);
+    DLAF_ASSERT_HEAVY(i_begin <= i_end, i_begin, i_end);
     return globalElementFromGlobalTileAndTileElement<rc>(i_end, 0) -
            globalElementFromGlobalTileAndTileElement<rc>(i_begin, 0);
   }
 
+  /// Returns the global element size of the region between global tile indices @p i_begin and @p i_end
+  /// (including) along the @p rc coordinate
+  template <Coord rc>
+  SizeType globalElementSizeFromGlobalTileIndexRange(SizeType i_begin, SizeType i_end) const noexcept {
+    DLAF_ASSERT_HEAVY(i_begin <= i_end, i_begin, i_end);
+    DLAF_ASSERT_HEAVY(0 <= i_begin && i_end < global_nr_tiles_.get<rc>(), i_begin, i_end);
+    SizeType n = size_.get<rc>();
+    SizeType nb = block_size_.get<rc>();
+    SizeType nbr = std::min(nb, n - i_end * nb);  // size of last tile along `rc`
+    return (i_end - i_begin) * nb + nbr;
+  }
+
+  /// Returns the local element size of the region between global tile indices @p i_begin and @p i_end
+  /// (including) along the @p rc coordinate
+  template <Coord rc>
+  SizeType localSizeFromGlobalTileIndexRange(SizeType i_begin, SizeType i_end) const noexcept {
+    DLAF_ASSERT_HEAVY(i_begin <= i_end, i_begin, i_end);
+    DLAF_ASSERT_HEAVY(0 <= i_begin && i_end < global_nr_tiles_.get<rc>(), i_begin, i_end);
+
+    SizeType i_loc_begin = nextLocalTileFromGlobalTile<rc>(i_begin);
+    SizeType i_loc_end = prevLocalTileFromGlobalTile<rc>(i_end);
+
+    if (i_loc_begin > i_loc_end)
+      return 0;
+
+    SizeType l = local_size_.get<rc>();
+    SizeType nb = block_size_.get<rc>();
+    SizeType nbr = std::min(nb, l - i_loc_end * nb);  // size of last local tile along `rc`
+
+    return (i_loc_end - i_loc_begin) * nb + nbr;
+  }
```

Comment:
The end index (similar to the end iterator) is NOT included in the range. If we start using the wrong notation in the distribution object it will be source of problems for sure.

Moreover, it is not clear to me the reason why in the tridiagonal solver we need the last index and we cannot simply use the end index and simplify things.
E.g.:
`globalElementSizeFromGlobalTileIndexRange(i_begin, i_last) = globalTileElementDistance(i_begin, i_end = i_last + 1)`
and the local version can be easily implemented using twice `nextLocalTileFromGlobalTile`
```suggestion
  /// ...
  template <Coord rc>
  SizeType localTileElementDistance(SizeType i_begin, SizeType i_end) const noexcept {
    DLAF_ASSERT_HEAVY(i_begin <= i_end, i_begin, i_end);
    // Note the second assert is already done by the following calls.

    SizeType i_loc_begin = nextLocalTileFromGlobalTile<rc>(i_begin);
    SizeType i_loc_last = nextLocalTileFromGlobalTile<rc>(i_end) - 1;

    if (i_loc_begin > i_loc_last)
      return 0;

    SizeType l = local_size_.get<rc>();
    SizeType nb = block_size_.get<rc>();
    SizeType nbr = std::min(nb, l - i_loc_last * nb);  // size of last local tile along `rc`

    return (i_loc_last - i_loc_begin) * nb + nbr;
  }
```
Maybe I'm too used to `for(i = start; i < end; ++i)`, but using the last index looks too similar to Fortran... 😉

Reply:
Great point! I opened an issue regarding the exclusion of the end index : https://github.com/eth-cscs/DLA-Future/issues/708 . Multiple algorithms are affected so I suggest we change it in a separate PR.

I have the addressed the remaining comments in this commit : https://github.com/eth-cscs/DLA-Future/pull/676/commits/b72c403a453739c38cdd97d778259c137605b5c2 . I removed all unneeded functions and resorted to the one you suggest, including calls to `nextLocalTileFromGlobalTile` where `prevLocalTileFromGlobalTile` was previously used.